### ISSUE TITLE
Promise not being resolved until shutdown?

### ISSUE THREAD

**Issue by timscaffidi** [guzzle/guzzle#1478]
Given the simple example below:

    $promise = $this->getClient()->postAsync(
        $uri,
        [
            'json' => $json_body,
            'headers' => [
                'X-Css-User' => $this->getApiUser()
            ]
        ]
    );

    $promise->then(function () {
        echo "promise resolved!!!\n";
    });

    $response = $promise->wait();

    echo "after wait\n";

I see the output:

    after wait
    promise resolved

This indicates that nothing is actually resolving the promise when calling wait(), since the 'after wait' message is printed first.

So when is the promise resolved?

Adding in a `var_export(debug_backtrace())` to the callback function in the `then()` call, I can see that the promise is finally being resolved due to the [register_shutdown_function call in TaskQueue.php](https://github.com/guzzle/promises/blob/8c54fb01b3b077f07f74f658eac1c558930cfbc3/src/TaskQueue.php#L26).

I would expect any registered callbacks to be called as soon as the response is available. Am I doing something wrong, or is this the intended behavior?

**Comment by jeskew** [comment]
There are two promises being created in your sample: one named `$promise` and one created by calling `$promise->then`, which always returns a new promise. Only the first promise is being explicitly waited on, so the second one is left unresolved until the TaskQueue's registered shutdown function is invoked.

**Comment by timscaffidi** [comment]
I see that there are two promises, however it seems counter intuitive to me that the function being passed into the then() call would not be called immediately when $promise is resolved. Semantically, the code reads to me "when this promise is resolved, then do this".

I guess my point is that the wait() call to the secondary promise should happen automatically if/when the parent promise is resolved.

This is how many other promises APIs work, for example in angular.js.

**Comment by jeskew** [comment]
`then` does not register a callback on the original promise; instead, it creates a new promise that is contingent on the first and throws a new task on the queue. 

[This is actually how Angular promises work as well](https://github.com/angular/angular.js/blob/master/src/ng/q.js#L312), though most JavaScript promise libraries use the language's global event loop to run through their task queues. Since PHP does not have a global event loop, you will need to wait on the second promise or [integrate with an external event loop](https://github.com/guzzle/promises#event-loop-integration).

**Comment by timscaffidi** [comment]
Thanks for the clarification, this makes sense now. I was not aware of the built in event loop in javascript to handle this.
I was able to get the expected behavior by adding:

    Promise\queue()->run();

After setting up the new promise.
